docs/grfn_spec.md

```diff
@@ -11,96 +20,189 @@ associated extracted comments, links to natural language text (the
+
+
+Identifiers: grounding, scopes, namespaces and gensyms
+------------------------------------------------------
+
+An identifier is a symbol used to uniquely identify a program element in code, where a *program element* is a 
+- variable (or constant)
+- function
+- type (class)
+
+More than one identifier can be used to denote the same program element, but an identifier can only be associated with one program element at a time.
+
+### Grounding
+
+Identifiers play a key role in connecting the model as implemented in source code to the target domain that it models. *Grounding* is the task of inferring what aspect of target domain a program element may correspond to, and identifiers, by their (base) name(s), their declaration and use (where they occur in code -- `scope` and `namespace`), and the doc and comment strings that occur around them, provide clues to what they are intended to refer to. For this reason, we need to associate with identifiers several pieces of information. This information will be collected during program analysis and associated with the identifier declaration:
+
+- "aliases": It is possible for multiple identifiers to be used to denote the same program element. How this is done differs across languages, according to scoping rules and assignment. Program analysis modules for each language (e.g., the Fortran `for2py` analyzer) will determine this. One general way to assign more than one identifier to the same program element is through a *simple equality assignment*, e.g.: `y = x` means that a new identifier, `y`, denotes the same program element that `x` does. A simple equality assignment just involve one identifier being equated with another, no other operations are applied; if other operations are applied (e.g., `y = x + 1`), then this is a *new* identifier as it does not represent the original value of `x` but a modification of it.
+
+    CHOICE: Do we declare each identifier separately, or combine them at program analysis time to the a single identifier with aliases.
+    
+    FOR NOW: We will only keep track of a single identifier (the first one encountered by program analysis) but associate any additional `aliases` as the names of any additional identifier introduced in code.
+	
+- "source\_references": To facilitate later grounding inference, we will store a reference to the location within the source code where an identifier is declared, using a `<source_code_reference>`:
+
+	```
+    <source_code_reference> := <string>
```

Comment:
I thought we were using `::=` to denote a new rule in BNF (as opposed to `:=`)? Is this not a rule statement?